js/plugins/conviva.js

```diff
@@ -0,0 +1,941 @@
+   */
+  var resetPlaybackState = function()
+  {
+    currentPlayhead = -1;
+    buffering = false;
+    inAdBreak = false;
+    contentComplete = false;
+  };
+
+  /**
+   * [Required Function] Clean up this plugin so the garbage collector can clear it out.
+   * @public
+   * @method ConvivaAnalyticsPlugin#destroy
+   */
+  this.destroy = function()
+  {
+    _framework = null;
+    resetPlaybackState();
+    clearLastSession();
+    if (convivaClient)
+    {
+      convivaClient.release();
+      convivaClient = null;
+    }
+
+    if (systemFactory)
+    {
+      systemFactory.release();
+      systemFactory = null;
+    }
+  };
+
+  var validSession = function()
+  {
+    return currentConvivaSessionKey !== Conviva.Client.NO_SESSION_KEY && currentConvivaSessionKey !== null;
+  };
+
+  /**
+   * Checks to see if the Conviva SDK is ready to accept tracking events.
+   * @private
+   * @method ConvivaAnalyticsPlugin#canTrack
+   */
+  var canTrack = function()
+  {
+    return playerStateManager && convivaClient && validSession();
+  };
+
+  /**
+   * Updates the Conviva PlayerStateManager of the latest player state.
+   * @private
+   * @method ConvivaAnalyticsPlugin#updatePlayerState
+   * @param {string} state the Conviva.PlayerStateManager.PlayerState to update
+   */
+  var updatePlayerState = function(state)
+  {
+    if (canTrack())
+    {
+      playerStateManager.setPlayerState(state);
+    }
+    else
+    {
+      OO.log("Conviva Plugin Error: trying to set player state when unable to with state " + state);
+    }
+  };
+
+  /**
+   * To be called when the main content has started playback.
+   * @private
+   * @method ConvivaAnalyticsPlugin#trackPlay
+   */
+  var trackPlay = function()
+  {
+    updatePlayerState(Conviva.PlayerStateManager.PlayerState.PLAYING);
+  };
+
+  /**
+   * To be called when the main content has paused.
+   * @private
+   * @method ConvivaAnalyticsPlugin#trackPause
+   */
+  var trackPause = function()
+  {
+    updatePlayerState(Conviva.PlayerStateManager.PlayerState.PAUSED);
+  };
+
+  /**
+   * To be called when the main content is not playing.
+   * @private
+   * @method ConvivaAnalyticsPlugin#trackStop
+   */
+  var trackStop = function()
+  {
+    updatePlayerState(Conviva.PlayerStateManager.PlayerState.STOPPED);
+  };
+
+  /**
+   * To be called when the video is paused due to buffering.
+   * @private
+   * @ethod ConvivaAnalyticsPlugin#trackBuffering
+   */
+  var trackBuffering = function()
+  {
+    updatePlayerState(Conviva.PlayerStateManager.PlayerState.BUFFERING);
+  };
+
+  /**
+   * To be called when an ad has started playing.
+   * @private
+   * @method ConvivaAnalyticsPlugin#trackAdStart
+   */
+  var trackAdStart = function()
+  {
+    if (canTrack())
+    {
+      var adPosition = null;
+      if (currentPlayhead <= 0)
+      {
+        OO.log("[Conviva-Ooyala] Playing preroll");
+        adPosition = Conviva.Client.AdPosition.PREROLL;
```

Comment:
The SDK should guarantee that Conviva.Client exists. We check if the SDK has loaded. This should be sufficient.